From 723a355b342f66629ae089c1cbd729f363405d03 Mon Sep 17 00:00:00 2001
From: Armin Kuster <akuster@mvista.com>
Date: Fri, 15 Jul 2016 17:19:17 -0700
Subject: [PATCH] arm: amlogic: Enable MMC driver on Odroid-C2From: Carlo
 Caione <carlo at endlessm.com>

Add support for the MMC on the Hardkernel Odroid-C2 board.

Signed-off-by: Carlo Caione <carlo at endlessm.com>
Signed-off-by: Armin Kuster <akuster@mvista.com>
---
 board/amlogic/odroid-c2/odroid-c2.c | 41 +++++++++++++++++++++++++++++++++++++
 configs/odroid-c2_defconfig         |  5 +++++
 include/configs/odroid-c2.h         |  7 +++++++
 3 files changed, 53 insertions(+)

Index: git/board/amlogic/odroid-c2/odroid-c2.c
===================================================================
--- git.orig/board/amlogic/odroid-c2/odroid-c2.c
+++ git/board/amlogic/odroid-c2/odroid-c2.c
@@ -8,6 +8,7 @@
 #include <asm/io.h>
 #include <asm/arch/gxbb.h>
 #include <asm/arch/sm.h>
+#include <asm/arch/sd_emmc.h>
 #include <dm/platdata.h>
 #include <phy.h>
 
@@ -65,3 +66,43 @@ int misc_init_r(void)
 
 	return 0;
 }
+
+#if defined(CONFIG_GENERIC_MMC) 
+
+static const struct meson_mmc_platdata gxbb_sd_platdata[] = {
+	{ .sd_emmc_reg = (struct meson_mmc_regs *)SD_EMMC_BASE_A },
+	{ .sd_emmc_reg = (struct meson_mmc_regs *)SD_EMMC_BASE_B },
+	{ .sd_emmc_reg = (struct meson_mmc_regs *)SD_EMMC_BASE_C },
+};
+
+U_BOOT_DEVICE(meson_mmc) = {
+	.name = "meson_mmc",
+	.platdata = &gxbb_sd_platdata[CONFIG_MMC_MESON_SD_PORT],
+};
+
+static void meson_mmc_pinmux_setup(unsigned int port)
+{
+	switch (port) {
+	case SDIO_PORT_A:
+		setbits_le32(GXBB_PINMUX(8), 0x3f);
+		break;
+	case SDIO_PORT_B:
+		setbits_le32(GXBB_PINMUX(2), 0x3f << 10);
+		break;
+	case SDIO_PORT_C:
+		clrbits_le32(GXBB_PINMUX(2), 0x1f << 22);
+		setbits_le32(GXBB_PINMUX(4), (0x3 << 18) | (3 << 30));
+		break;
+	default:
+		printf("meson: invalid MMC port %d for pinmux setup\n", port);
+		break;
+	}
+}
+
+int board_mmc_init(bd_t *bis)
+{
+	meson_mmc_pinmux_setup(CONFIG_MMC_MESON_SD_PORT);
+
+	return 0;
+}
+#endif
Index: git/configs/odroid-c2_defconfig
===================================================================
--- git.orig/configs/odroid-c2_defconfig
+++ git/configs/odroid-c2_defconfig
@@ -8,11 +8,16 @@ CONFIG_HUSH_PARSER=y
 # CONFIG_CMD_IMI is not set
 # CONFIG_CMD_IMLS is not set
 # CONFIG_CMD_LOADS is not set
+CONFIG_CMD_MMC=y
 # CONFIG_CMD_FPGA is not set
 # CONFIG_CMD_SOURCE is not set
 # CONFIG_CMD_SETEXPR is not set
+CONFIG_CMD_EXT4=y
+CONFIG_CMD_FAT=y
+CONFIG_CMD_FS_GENERIC=y
 CONFIG_OF_CONTROL=y
 CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_DM_MMC=y
 CONFIG_DM_ETH=y
 CONFIG_DEBUG_UART=y
 CONFIG_DEBUG_UART_MESON=y
Index: git/include/configs/odroid-c2.h
===================================================================
--- git.orig/include/configs/odroid-c2.h
+++ git/include/configs/odroid-c2.h
@@ -14,6 +14,13 @@
 #define CONFIG_CONS_INDEX		0
 #define CONFIG_BAUDRATE			115200
 
+#ifdef CONFIG_DM_MMC
+#define CONFIG_MMC
+#define CONFIG_GENERIC_MMC
+#define CONFIG_MMC_MESON
+#define CONFIG_MMC_MESON_SD_PORT	1
+#endif
+
 #include <configs/meson-gxbb-common.h>
 
 #endif /* __CONFIG_H */
